File: src/test/java/org/hejwo/java/algorithms/sorting/SelectionSortTest.java
package org.hejwo.java.algorithms.sorting;

import java.util.List;
import java.util.StringJoiner;

import org.hejwo.java.algorithms.sorting.selectionsort.SelectionSort;
import org.junit.Test;

import com.google.common.collect.Lists;

import static org.assertj.core.api.Assertions.assertThat;

public class SelectionSortTest {

    @Test
    public void sort() {
        List<Integer> unsorted = Lists.newArrayList(8, 7, 1, 2, 4, 3, 6, 5);

        List<Integer> sorted = SelectionSort.sort(unsorted);

        assertThat(sorted).containsExactly(1, 2, 3, 4, 5, 6, 7, 8);
    }

    @Test
    public void mergeSort() {
        List<Integer> unsorted = Lists.newArrayList(32, 1, 2, 4, 3, 6, 7, 8, 5);

        List<Integer> sorted = SelectionSort.sort(unsorted);
        printResult(sorted);

        assertThat(sorted).containsExactly(1, 2, 3, 4, 5, 6, 7, 8, 32);
    }

    @Test
    public void sort_shouldHandleSimpleCase() {
        List<Integer> unsorted = Lists.newArrayList(8, 7, 1, 2, 4, 3, 6, 5);

        List<Integer> sorted = SelectionSort.sort(unsorted);

        assertThat(sorted).containsExactly(1, 2, 3, 4, 5, 6, 7, 8);
    }

    @Test
    public void sort_shouldReturnSorted_whenAlreadySorted() {
        List<Integer> unsorted = Lists.newArrayList(1, 2, 3, 4, 5, 6, 7, 8);

        List<Integer> sorted = SelectionSort.sort(unsorted);

        assertThat(sorted).containsExactly(1, 2, 3, 4, 5, 6, 7, 8);
    }

    @Test
    public void sort_shouldReturnSameOne_whenEmpty() {
        List<Integer> unsorted = Lists.newArrayList();

        List<Integer> sorted = SelectionSort.sort(unsorted);

        assertThat(sorted).isEqualTo(unsorted);
    }

    @Test
    public void sort_shouldReturnSameOne_whenSingleElement() {
        List<Integer> unsorted = Lists.newArrayList(124);

        List<Integer> sorted = SelectionSort.sort(unsorted);

        assertThat(sorted).isEqualTo(unsorted);
    }

    @Test
    public void sort_shouldReturnSameOne_whenNull() {
        List<Integer> unsorted = null;

        List<Integer> sorted = SelectionSort.sort(unsorted);

        assertThat(sorted).isEqualTo(unsorted);
    }

    private void printResult(List<Integer> sorted) {
        StringJoiner stringJoiner = new StringJoiner(",");
        sorted.forEach(val -> stringJoiner.add(val.toString()));
        System.out.println(stringJoiner.toString());
    }
    

}

File: src/test/scala/org/hejwo/scala/algorithms/sorting/selectionsort/SelectionSortTest.scala
package org.hejwo.scala.algorithms.sorting.selectionsort

import org.scalatest.{FlatSpec, FunSpec, FunSuite}

class SelectionSortTest extends FunSuite {

  test("empty list should return empty list") {
    val unsorted = List()
    val sorted = SelectionSort.sort(unsorted)

    assert(sorted.size === 0)
    assert(sorted === unsorted)
  }

  test("null list should return null list") {
    val unsorted = null
    val sorted = SelectionSort.sort(unsorted)

    assert(sorted === null)
  }

}

class SelectionSortTest2 extends FlatSpec {

  import SelectionSort.sort

  "empty list" should "return empty list" in {
    val unsorted = List()
    val sorted = sort(unsorted)

    assert(sorted.size === 0)
    assert(sorted === unsorted)
  }

  "null list" should "return null list" in {
    val unsorted = null
    val sorted = sort(unsorted)

    assert(sorted === null)
  }

  "unsorted list without duplicates" should "be sorted" in {
    val unsorted = List(3, 1, 2, 4, 5)
    val sorted = sort(unsorted)

    assert(sorted === List(1, 2, 3, 4, 5))
  }

  "unsorted with duplicates" should "be sorted" in {
    val unsorted = List(3, 1, 4, 2, 4, 5)
    val sorted = sort(unsorted)

    assert(sorted === List(1, 2, 3, 4, 4, 5))
  }

  "sorted list without duplicates" should "return sorted" in {
    val unsorted = List(1, 2, 3, 4, 5)
    val sorted = sort(unsorted)

    assert(sorted === List(1, 2, 3, 4, 5))
  }

}

class SelectionSortTest3 extends FunSpec {

}
